lib/tpm2_errata.c

```diff
@@ -0,0 +1,212 @@
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+#include <stdarg.h>
+#include <stdbool.h>
+
+#include "log.h"
+#include "tpm2_errata.h"
+#include "tpm2_util.h"
+
+struct tpm2_errata_desc {
+    UINT32 spec_rev;
+    UINT32 errata_ver;
+    void (*fixup) (struct tpm2_errata_desc *errata, va_list *ap);
+    const char *name;
+};
+
+/* Published errata information */
+struct tpm2_errata_info {
+    UINT32 spec_rev;
+    UINT32 errata_ver;
+    UINT32 year;
+    UINT32 day_of_year;
+} known_errata_info[] = {
+    /* Specification Revision 1.16 October 30, 2014 */
+    { 116, 000, 2014, 303 },
+    /* Errata Version 1.2 February 16, 2015 */
+    { 116, 120, 2015,   4 },
+    /* Errata Version 1.3 June 16, 2015 */
+    { 116, 130, 2015, 167 },
+    /* Errata Version 1.4 January 15, 2016 */
+    { 116, 140, 2016,  15 },
+    /* Errata Version 1.5 September 21, 2016 */
+    { 116, 150, 2016, 265 },
+    /* Specification Revision 1.38 September 29, 2016 */
+    { 138, 000, 2016, 273 },
+    /* Errata Version 1.0 January 16, 2017 */
+    { 138, 100, 2017,  16 },
+    /* Errata Version 1.1 March 2, 2017 */
+    { 138, 110, 2017,  61 },
+}, *this_errata_info;
+
+static void
+fixup_sign_decrypt_attribute_encoding(struct tpm2_errata_desc *errata,
+                                      va_list *ap) {
+
+    TPMA_OBJECT *attrs = va_arg(*ap, TPMA_OBJECT *);
+printf("applied\n");
+    attrs->sign = 0;
+
+    LOG_INFO("Errata %s applied\n", errata->name);
+}
+
+/*
+ * Note that each record contains the first errata version with the
+ * corresponding errata correction.
+ */
+static struct tpm2_errata_desc sign_decrypt_attribute_encoding = {
+    .spec_rev = 116,
+    .errata_ver = 120,
+    .fixup = fixup_sign_decrypt_attribute_encoding,
+    .name = "Sign/decrypt attribute encoding",
+};
+
+/* Add the new errata descriptor here */
+
+static struct tpm2_errata_desc *errata_desc_list[] = {
+    [SPEC_116_ERRATA_2_7] = &sign_decrypt_attribute_encoding,
+    /* Append the new errata descriptor here */
+};
+
+static bool errata_match(struct tpm2_errata_desc *errata) {
+
+    return errata->spec_rev >= this_errata_info->spec_rev &&
+           errata->errata_ver > this_errata_info->errata_ver;
+}
+
+static bool errata_query(int nr, struct tpm2_errata_desc **errata) {
+
+    if (nr < 0 || (size_t)nr >= sizeof(errata_desc_list) / sizeof(*errata)) {
+        LOG_ERR("Invalid errata requested: %d\n", nr);
+        return false;
+    }
+
+    if (!this_errata_info) {
+        LOG_ERR("Unknow TPM_SPEC\n");
+        return false;
+    }
+
+    *errata = errata_desc_list[nr];
+
+    return errata_match(*errata);
+}
+
+bool tpm2_errata_fixup(int nr, ...) {
+
+    struct tpm2_errata_desc *errata;
+    va_list ap;
+
+    if (errata_query(nr, &errata) == false)
+        return false;
+
+    va_start(ap, nr);
+    errata->fixup(errata, &ap);
+    va_end(ap);
+
+    return true;
+}
```

Comment:
Can you please add the commits that adds the errata queries for the tpm2 tools? I think that would make more easier to understand how these functions will be used.

Reply:
No problem. Actually I have planed to do it soon. That tool would give the full errata list for a running TPM.